File: Lecture30/random_test.py
import math
import random
import time

poor_seed = 123                         # maintain seed for poor generator
def poor():                             # example from Numerical Recipes
    global poor_seed
    seed = poor_seed
    IM = 6075 ; IA = 106 ; IC = 1283
    seed = (IA * seed + IC) % IM        # linear congruential algorithm
    poor_seed = seed
    return seed / float(IM)

park_miller_seed = 1                    # maintain seed for Park-Miller
def park_miller():                      # Park-Miller generator
    global park_miller_seed
    seed = park_miller_seed
    IA = 16807 ; IC = 2147483647 ; IQ = 127773 ; IR = 2836
    h = seed // IQ
    l = seed % IQ;
    seed = IA * l - IR * h
    if seed <= 0:
        seed += IC
    park_miller_seed = seed
    return seed / float(IC)

def test(generator, name, remark):
    print " " + name + "() " + remark

    print " checking period ..."
    start = generator()
    max_steps = 1000000
    for i in range(max_steps):
        if generator() == start:
            print " repeats after " + repr(i) + " steps"
            break
    if i+1 >= max_steps:
        print " period larger than " + repr(max_steps)

    start_time = time.clock()
    bins = 10000
    print " binning " + repr(max_steps) + " tries in " + repr(bins) + " bins"
    bin = [ 0.0 ] * bins
    for i in range(max_steps):
        b = int(generator() * bins)
        if b >= 0 and b < bins:
            bin[b] += 1
    chisqr = 0.0
    expect = max_steps / float(bins)
    for b in range(bins):
        diff = bin[b] - expect
        chisqr += diff**2
    chisqr /= expect
    cpu_time = time.clock() - start_time
    print " chi-squar/d.o.f. = " + repr(chisqr/(bins-1.0))
    print " CPU time = " + repr(cpu_time) + " seconds"

    file_name = name + ".data"
    print " writing eyeball test data to " + file_name
    file = open(file_name, "w")
    for i in range(10000):
        file.write(repr(generator()) + "\t" + repr(generator()) + "\n")
    file.close()
    print

print " Testing random number generators()"
print " ----------------------------------"
test(poor, "poor", "from Numerical Recipes period 6,075")
test(park_miller, "park_miller", "Park-Miller multiplicative generator")
test(random.random, "random.random", "from Python standard library")

File: Lecture30/random_test.cpp
#include "cptstd.hpp"
#include "random.hpp"
using namespace cpt;

int poor_seed = 123;                    // maintain seed for poor generator
double poor() {                         // example from numerical recipes
    const int IM = 6075, IA = 106, IC = 1283;
    int seed = poor_seed;
    seed = (IA * seed + IC) % IM;       // linear congruential algorithm
    poor_seed = seed;
    return seed / double(IM);           // make range 0.0 ... 1.0
}

int park_miller_seed = 1;
double park_miller() {                  // Park-Miller generator
    const int IA = 16807, IC = 2147483647, IQ = 127773, IR = 2836;
    int seed = park_miller_seed;
    int h = seed / IQ;
    int l = seed % IQ;
    seed = IA * l - IR * h;
    if (seed <= 0)
        seed += IC;
    park_miller_seed = seed;
    return seed / double(IC);
}

int std_rand_seed;
double std_rand() {                     // C++ standard library generator
    std_rand_seed = rand();
    return std_rand_seed / (RAND_MAX + 1.0);
}

Random rg;
double xorshift() {
    double r = rg.rand();
    return r;
}

double mersenne() {
    double r = rg.rand();
    return r;
}

void test(double generator(), string name, string remark) {
    cout << " " << name << "() -- " << remark << endl;

    cout << " checking period ... " << flush;
    double start = generator();
    long int i, max_steps = 10000000L;
    for (i = 0L; i < max_steps; i++) {
        if (generator() == start) {
            cout << " repeats after "  << i << " steps\n";
            break;
        }
    }
    if (i >= max_steps)
        cout << " period larger than " << max_steps << endl;

    clock_t start_time = clock();
    const int BINS = 10000;
    cout << " binning " << max_steps << " tries in "
         << BINS << " bins ..." << flush;
    int bin[BINS];
    for (int b = 0; b < BINS; b++)
        bin[b] = 0;
    for (i = 0L; i < max_steps; i++) {
        int b = int(generator() * BINS);
        if (b == BINS)
          --b;
        ++bin[b];
    }
    double chiSqr = 0;
    double expect = max_steps / double(BINS);
    for (int b = 0; b < BINS; b++) {
        double diff = bin[b] - expect;
        chiSqr += diff * diff;
    }
    chiSqr /= expect;
    cout << " chi-square/d.o.f. = " << chiSqr/(BINS-1) << endl;
    cout << " CPU time = " << (clock() - start_time) / double(CLOCKS_PER_SEC)
         << " seconds\n";

    string fileName = name + ".data";
    cout << " writing eyeball test data to " << fileName << endl;
    ofstream dataFile(fileName.c_str());
    for (int i = 0; i < 10000; i++)
        dataFile << generator() << '\t' << generator() << '\n';
    dataFile.close();
    cout << endl;
}

int main() {
    cout << " Testing random number generators\n"
         << " --------------------------------\n\n";
    test(poor, "poor", "from Numerical Recipes period 6,075");
    test(park_miller, "park_miller", "Park-Miller multiplicative generator");
    test(std_rand, "std_rand", "rand() from C++ Standard Library");
    rg.set_xorshift();
    test(xorshift, "xorshift", "Marsaglia Xorshift generator from random.hpp");
    rg.set_mersenne_twister();
    test(mersenne, "mersenne", "Mersenne Twister generator");
}
